### PR TITLE
[WIP] Pr 1259 event invitation not emailed to banned

### PR DESCRIPTION
**Pull request by notapatch**
This is my work so far for issue [#1259](
https://github.com/codebar/planner/issues/1259) - I thought I'd keep you in the loop as it was high priority.

This work removes banned members at the scope level. Proving a negative is hard so for all cases I had a test that allowed a banned member and did another test with the minimum change to reject the banned member.

I was surprised that `in_group` filtered out the banned members rather than call it `allowed_in_group` or an `allowed` scope but I thought, well, I can't think of any reason to allow any banned member appearing in invitations so the above changes had the intention of making it impossible to get a banned member in an invitation scope.

- ea60cfa  - test to confirm `in_group` removes banned members
- 72763d3 - test to confirm `attending_meeting` removes banned members
- 48a8240 - invitable concern was missing any testing - this tested the current code with the shared_example
- b0924be - Changes invitable concern to remove banned members plus testing

These changes will remove all banned members from InvitationManager's lists (because they end up calling the above banned cleared scopes). The exception is `retrieve_and_notify_waitlisted(role:)` which I will look at if the work so far is OK.

When `retrieve_and_notify_waitlised` is complete then there can be no banned members getting an invitation. However, the issue expressly wanted tests for each of the methods in the WorshopInvitationManagerConcern and that would be the next step of the work.

### CODE REVIEW

File: app/presenters/workshop_presenter.rb [hunk 1]
```diff
@@ -96,7 +96,9 @@ def pairing_csv
   private
 
   def students_checklist
-    model.attending_students.order('note asc').each_with_index.map do |a, pos|
+    # TODO: review if it needs this removed scope or not
+    #       Not going to worry unless the rest of the code is OKd
```
Comment (despo) [inline app/presenters/workshop_presenter.rb:100]: TODO can be removed as this does not require `not_banned` scope to be used. Ideally members should be removed from all RSVPed events as soon as they are banned, but even if this is missed, organisers would be able to see they are banned and they wouldnt be able to RSVP to any other future events so that is sufficient.

File: spec/models/event_spec.rb [hunk 2]
```diff
@@ -2,6 +2,7 @@
 
 RSpec.describe Event, type: :model  do
   subject(:event) { Fabricate(:event) }
+  include_examples "Invitable", :invitation, :event
```
Comment (despo) [inline spec/models/event_spec.rb:5]: 👍

File: spec/models/meeting_spec.rb [hunk 3]
```diff
@@ -1,6 +1,8 @@
 require 'spec_helper'
 
 RSpec.describe Meeting, type: :model  do
+  include_examples "Invitable", :meeting_invitation, :meeting
+
```
Comment (despo) [inline spec/models/meeting_spec.rb:5]: 👍

File: spec/support/shared_examples/behaves_like_invitable.rb [hunk 4]
```diff
@@ -0,0 +1,95 @@
+# shared examples for Invitable
+# Example usage:
+#  include_examples "Invitable", :workship_invitation, :workshop
+#
```
Comment (despo) [inline spec/support/shared_examples/behaves_like_invitable.rb:4]: This looks great, but can we remove the comment please. I think documentation about shared examples is easily available and we don't need comments in the code.

File: spec/support/shared_examples/behaves_like_invitable.rb [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+RSpec.shared_examples 'Invitable' do |invitation_type, invitable_type|
+  let(:invitable) { Fabricate(invitable_type) }
+
+  context '#attendances' do
+    it 'permits accepted' do
+      invitation = Fabricate(invitation_type,
+                             invitable_type => invitable,
```
Comment (despo) [inline spec/support/shared_examples/behaves_like_invitable.rb:11]: can this be updated to `invitable_type: invitable` please

File: spec/support/shared_examples/behaves_like_invitable.rb [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+                             attending: true)
+
+      expect(invitable.reload.attendances).to include(invitation)
+    end
+
+    it 'rejects un-accepted' do
+      invitation = Fabricate(invitation_type,
+                             invitable_type => invitable,
```
Comment (despo) [inline spec/support/shared_examples/behaves_like_invitable.rb:19]: can this be updated to `invitable_type: invitable` please

File: spec/support/shared_examples/behaves_like_invitable.rb [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+  context '#attendances' do
+    it 'permits accepted' do
+      invitation = Fabricate(invitation_type,
+                             invitable_type => invitable,
+                             attending: true)
+
+      expect(invitable.reload.attendances).to include(invitation)
+    end
+
+    it 'rejects un-accepted' do
```
Comment (despo) [inline spec/support/shared_examples/behaves_like_invitable.rb:17]: Can we change this to read non accepted. I find un-accepted a bit confusing.

File: spec/support/shared_examples/behaves_like_invitable.rb [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                             invitable_type => invitable,
+                             attending: false)
+
+      expect(invitable.reload.attendances).to_not include(invitation)
+    end
+
+    it 'rejects banned accepted' do
+      invitation = Fabricate(invitation_type,
+                             member: Fabricate(:banned_member),
+                             invitable_type => invitable,
```
Comment (despo) [inline spec/support/shared_examples/behaves_like_invitable.rb:28]: same as above

File: spec/support/shared_examples/behaves_like_invitable.rb [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+                                        invitable_type => invitable,
+                                        attending: true)
+
+      expect(invitable.reload.attending_students).to include(invitation_to_student)
+    end
+
+    it 'rejects un-attending students' do
+      invitation_to_student = Fabricate(invitation_type,
+                                        role: 'Student',
+                                        invitable_type => invitable,
```
Comment (despo) [inline spec/support/shared_examples/behaves_like_invitable.rb:48]: same as above

File: spec/support/shared_examples/behaves_like_invitable.rb [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+
+      expect(invitable.reload.attendances).to_not include(invitation)
+    end
+  end
+
+  context '#attending_students' do
+    it 'accepts attending students' do
+      invitation_to_student = Fabricate(invitation_type,
+                                        role: 'Student',
+                                        invitable_type => invitable,
```
Comment (despo) [inline spec/support/shared_examples/behaves_like_invitable.rb:39]: same as above

File: spec/support/shared_examples/behaves_like_invitable.rb [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+                                        attending: false)
+
+      expect(invitable.reload.attending_students).to_not include(invitation_to_student)
+    end
+
+    it 'rejects banned attending students' do
+      invitation_to_banned_student = Fabricate(invitation_type,
+                                        member: Fabricate(:banned_member),
+                                        role: 'Student',
+                                        invitable_type => invitable,
```
Comment (despo) [inline spec/support/shared_examples/behaves_like_invitable.rb:58]: same as above

File: spec/support/shared_examples/behaves_like_invitable.rb [hunk 12]
```diff
@@ -0,0 +1,95 @@
...
+                                        # byebug
+      expect(invitable.reload.attending_students).to_not include(invitation_to_banned_student)
+    end
+  end
+
+  context '#attending_coaches' do
+    it 'accepts attending coaches' do
+      invitation_to_coach = Fabricate(invitation_type,
+                                      role: 'Coach',
+                                      invitable_type => invitable,
```
Comment (despo) [inline spec/support/shared_examples/behaves_like_invitable.rb:70]: same as above

File: spec/support/shared_examples/behaves_like_invitable.rb [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+                                      attending: false)
+
+      expect(invitable.reload.attending_coaches).to_not include(invitation_to_coach)
+    end
+
+    it 'rejects banned attending coaches' do
+      invitation_to_banned_student = Fabricate(invitation_type,
+                                        member: Fabricate(:banned_member),
+                                        role: 'Student',
+                                        invitable_type => invitable,
```
Comment (despo) [inline spec/support/shared_examples/behaves_like_invitable.rb:89]: same as above

File: spec/support/shared_examples/behaves_like_invitable.rb [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+                                      invitable_type => invitable,
+                                      attending: true)
+
+      expect(invitable.reload.attending_coaches).to include(invitation_to_coach)
+    end
+
+    it 'rejects un-attending coaches' do
+      invitation_to_coach = Fabricate(invitation_type,
+                                      role: 'Coach',
+                                      invitable_type => invitable,
```
Comment (despo) [inline spec/support/shared_examples/behaves_like_invitable.rb:79]: same as above

File: app/presenters/workshop_presenter.rb (hunk 1 above)
Comment (notapatch) [reply app/presenters/workshop_presenter.rb:100]: `attending_students` along with all the other Invitable methods will remove banned Members. 

The TODO was to remind me that `.order('note asc')` was breaking the code. I don't actually know how it works now because there is already an ordering on attending_students [`.order('members.name, members.surname')`]  so a second in the chain order is ignored anyway. 

```
# current situation before I did anything was in effect:
.order('members.name, members.surname').order('note asc')
```


I will remove `order('note asc') `along with the TODO.

File: spec/support/shared_examples/behaves_like_invitable.rb (hunk 4 above)
Comment (notapatch) [reply spec/support/shared_examples/behaves_like_invitable.rb:4]: done!

File: app/presenters/workshop_presenter.rb (hunk 1 above)
Comment (despo) [reply app/presenters/workshop_presenter.rb:100]: I saw the PR on Slack and completely missed that it was WIP when I added my comments last night.

If anything, I think it would be more beneficial to order the attendees by RSVP time so that it gives a better idea to organisers about when someone RSVPed without having to necessarily check all logged RSVP times.

File: spec/support/shared_examples/behaves_like_invitable.rb (hunk 5 above)
Comment (notapatch) [reply spec/support/shared_examples/behaves_like_invitable.rb:11]: I did initially try JSON style/Ruby 1.9 hash style and I got the following error.

```ruby
  Failure/Error:
       invitation = Fabricate(invitation_type,
                              member: Fabricate(:banned_member),
                              invitable_type: invitable,
                              attending: true)
     
     ActiveRecord::UnknownAttributeError:
       unknown attribute 'invitable_type' for WorkshopInvitation.
     Shared Example Group: "Invitable" called from ./spec/models/workshop_spec.rb:5
     # ./spec/support/shared_examples/behaves_like_invitable.rb:22:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:80:in `block (2 levels) in <top (required)>'
     # ------------------
     # --- Caused by: ---
     # NoMethodError:
     #   undefined method `invitable_type=' for #<WorkshopInvitation:0x00007fa175f84eb8>
     #   ./spec/support/shared_examples/behaves_like_invitable.rb:22:in `block (3 levels) in <top (required)>'
```

I've done some reading around to try and understand this as best as I can (see references below). I haven't found anything authoritative [the following quote was the most interesting](https://stackoverflow.com/questions/8675206/is-there-any-difference-between-the-key-value-and-key-value-hash-no).

> The JavaScript style (key: value) is only useful if all of your Hash keys are "simple" symbols (more or less something that matches /\A[a-z_]\w*\z/i, AFAIK the parser uses its label pattern for these keys).

So, with the JSON style/Ruby 1.9 hash style is good for literal symbols - the colon at the end means it can convert it into symbol and use it as a hash key. In this case invitable_type is **not** a literal symbol it's a variable that holds a symbol, so I believe, it can't be converted into a symbol by the JSON style hash key implicitly - it thinks it a method and gives the above error.

There _may_ be ways around this and I've tried weird conversions that would look way worse the than hash rocket without success. 

Anyway, this is what I've tried. Do you know of a solution or would it be acceptable to leave it as a hash rocket?

### References
[Ruby docs Hash](https://ruby-doc.org/core-2.7.0/Hash.html)
[How to access a symbol hash key using a variable in Ruby](https://stackoverflow.com/questions/25537934/how-to-access-a-symbol-hash-key-using-a-variable-in-ruby)
[Is there any difference between the `:key => “value”` and `key: “value”` hash notations?](https://stackoverflow.com/questions/8675206/is-there-any-difference-between-the-key-value-and-key-value-hash-no)

File: app/presenters/workshop_presenter.rb (hunk 1 above)
Comment (notapatch) [reply app/presenters/workshop_presenter.rb:100]: It sounds like slack gives you the PR name and not the Github title - not sure what to do about that. 

It's definitely fixable. The ordering could be removed from the invitable#attending_students and invitable#attending_coaches and the order explicitly applied by having a `by_name` `by_rsvp_time` It looks like it would affect about 10 places. 🤷‍♂️

File: spec/support/shared_examples/behaves_like_invitable.rb (hunk 5 above)
Comment (notapatch) [reply spec/support/shared_examples/behaves_like_invitable.rb:11]: @despo I pressed the wrong button. I pressed resolved when I meant to press comment. (Github messaging isn't my strong point as I'm sure matyikriszta would attest to).

Comment (despo) [reply spec/support/shared_examples/behaves_like_invitable.rb:11]: I don't see it as resolved :)

However I'm happy with this explanation and will set all `key: value` comments I made to resolved.

File: app/presenters/workshop_presenter.rb (hunk 1 above)
Comment (despo) [reply app/presenters/workshop_presenter.rb:100]: @notapatch ah ok. Sounds like something for a new PR then so let's ignore my comment about changing the ordering here and I'll open a new issue.

Comment (despo) [reply app/presenters/workshop_presenter.rb:100]: https://github.com/codebar/planner/issues/1275

File: spec/support/shared_examples/behaves_like_invitable.rb (hunk 7 above)
Comment (notapatch) [reply spec/support/shared_examples/behaves_like_invitable.rb]: done! changed all un- to non-  I think the test pairings looked better 👍

**Comment by despo** [comment]
@notapatch 💯

**Comment by notapatch** [comment]
Haha, there's a couple more changes I was going to do. Shall I open another PR or leave it here?

The situation is that Invitable will filter out banned members - so that should mean InvitationManager will have removed banned members with the exception of `retrieve_and_notify_waitlisted`.

If I had continued the TODO was:
1. Move `attendee?` from Workshop into Invitable as it depended on Invitable methods
   -  I thought it would help the focus of Workshop.

2. Look at Waitlist and make sure no banned members could get into it.

Not pushing - as long as you know Waitlist hasn't been secured. I'm happy.

**Comment by despo** [comment]
@notapatch 🤦  you are welcome to do 1 on a new PR. As for 2, I wouldn't. Banned members should not be able to perform any actions on invitations unless they have RSVPed and I think a better way to tackle this would be to update `Admin/BanController` to instead remove a banned member from any upcoming events by checking if they have any upcoming attending invitations (for all types of events).
